crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -699,59 +713,32 @@ def nodes(component = "all")
           substep = :compute_nodes
         end
 
-        ::Crowbar::UpgradeStatus.new.save_substep(substep, :running)
+        status.save_substep(substep, :running)
 
         # decide what needs to be upgraded
         case component
-        # Upgrade everything
         when "all"
-          if substep == :ceph_nodes
-            join_ceph_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-            substep = :controller_nodes
-          end
-          if substep == :controller_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :running)
-            upgrade_controller_clusters
-            upgrade_non_compute_nodes
-            prepare_all_compute_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-            substep = :compute_nodes
-          end
-          if substep == :compute_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :running)
-            upgrade_all_compute_nodes
-            finalize_nodes_upgrade
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-            ::Crowbar::UpgradeStatus.new.end_step
-          end
-        # Upgrade controller clusters only
+          # Upgrade everything
+          do_controllers_substep(substep)
+          substep = :compute_nodes
+          upgrade_all_compute_nodes
         when "controllers"
-          if substep == :ceph_nodes
-            join_ceph_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-            substep = :controller_nodes
-          end
-          if substep == :controller_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :running)
-            upgrade_controller_clusters
-            upgrade_non_compute_nodes
-            prepare_all_compute_nodes
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-          end
-        # Upgrade given compute node
+          # Upgrade controller clusters only
+          do_controllers_substep(substep)
         else
+          # Upgrade given compute node
           upgrade_one_compute_node component
           ::Crowbar::UpgradeStatus.new.save_substep(substep, :node_finished)
+        end
 
-          # if we're done with the last one, finalize
-          progress = ::Crowbar::UpgradeStatus.new.progress
-          if progress[:remaining_nodes].zero?
-            ::Crowbar::UpgradeStatus.new.save_substep(substep, :finished)
-            finalize_nodes_upgrade
-            ::Crowbar::UpgradeStatus.new.end_step
-          end
+        # if all nodes are upgraded, cleanup and end the whole step
+        status = ::Crowbar::UpgradeStatus.new
+        if status.progress[:remaining_nodes].zero?
+          status.save_substep(substep, :finished)
+          finalize_nodes_upgrade
+          status.end_step
         end
+        true
```

Comment:
Good point, it wasn't there before. I tried to add it to appease my failing testsuite - I still don't see why it is suddenly needed.

I'll need to investigate then...